### PR TITLE
Benchmark metric ingest

### PR DESCRIPTION
**Pull request by niksajakovljevic**
Read Prometheus data files and generate ingest load.
Using native Prometheus data format makes is easy to swap input data
which can be handy when debugging ingest performance.

Being able to run ingest benchmarks locally will enable faster feedback loop while working on performance improvements.

What differentiate `remote-write benchmarker` and this tool:
- runs in tests (meaning we could easily use it to compare benchmark runs)
- does not care about HTTP endpoint (only tests ingest code)
- can parallel load metrics in shards but does not replicate what Prometheus queue manager is doing (the focus here is not to replicate real-world scenario but do the least of what's needed to put ingestion on stress)

### CODE REVIEW

File: pkg/tests/testsupport/metric_loader.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	i.curBlockIdx++
+	if err := i.loadBlockSamples(); err != nil {
+		panic(err)
+	}
+	i.curSampleIdx = 0
+	return true
+}
+
+func (it *promIterator) loadBlockSamples() error {
+	log.Info("loading blocks, total samples", it.blocks[it.curBlockIdx].Meta().Stats.NumSamples, "series", it.blocks[it.curBlockIdx].Meta().Stats.NumSeries)
```
Comment (paulfantom) [inline pkg/tests/testsupport/metric_loader.go:82]:
```suggestion
	log.Infof("loading blocks, total samples: %v, series: %v",
			it.blocks[it.curBlockIdx].Meta().Stats.NumSamples,
			it.blocks[it.curBlockIdx].Meta().Stats.NumSeries)
```
nit: a bit easier to read in code review view.

**Review by paulfantom — COMMENTED** [review]
I did a quick review pass of this and got few minor issues:
- there are some issues with library imports
- Could we use git-lfs to store binary file (prometheus-data.tar.gz)?

File: pkg/tests/testsupport/metric_loader.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+		"series", it.blocks[it.curBlockIdx].Meta().Stats.NumSeries)
+	it.blockSamples = make([]*BlockSample, it.blocks[it.curBlockIdx].Meta().Stats.NumSamples)
+	it.labelsCache = make(map[uint64]labels.Labels, it.blocks[it.curBlockIdx].Meta().Stats.NumSeries)
+	querier, err := tsdb.NewBlockQuerier(it.blocks[it.curBlockIdx], math.MinInt64, math.MaxInt64)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := querier.Close(); err != nil {
+			log.Error(err)
```
Comment (antekresic) [inline pkg/tests/testsupport/metric_loader.go:92]:
```suggestion
			log.Error("err", err)
```

File: pkg/tests/testsupport/metric_loader.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+				}
+			}
+			if len(req.Timeseries) > 0 {
+				// flush leftovers
+				if _, _, err := ingest(context.Background(), &req); err != nil {
+					log.Error(err)
+				}
+			}
+		}(i)
+		wg.Add(1)
```
Comment (antekresic) [inline pkg/tests/testsupport/metric_loader.go:242]: This line should go before goroutine. Its idiomatic to show the goroutine which will be responsible for sending the done signal.

File: pkg/tests/testsupport/metric_loader.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+	shards    []chan prompb.TimeSeries
+	batchSize int
+	rate      *rate.Limiter
+}
+
+// NewSampleIngestor creates sampleIngestor responsible for splitting samples into shards and sending them down the ingestion path
+// zero sampleRate means no limits
+func NewSampleIngestor(shards int, samples PromIterator, batchSize int, sampleRate int) *sampleIngestor {
+	var r *rate.Limiter
+	if sampleRate != 0 {
```
Comment (antekresic) [inline pkg/tests/testsupport/metric_loader.go:174]: We could use `rate.Inf` for `sampleRate == 0`. Would simplify the rate limiting logic (no need for checking nils).

File: pkg/tests/testsupport/metric_loader.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+		Value:     blockSample.val,
+		Timestamp: blockSample.ts,
+	}
+	ts := prompb.TimeSeries{
+		Labels:               protoLabels,
+		Samples:              []prompb.Sample{sample},
+		Exemplars:            []prompb.Exemplar{},
+		XXX_NoUnkeyedLiteral: struct{}{},
+		XXX_unrecognized:     []byte{},
+		XXX_sizecache:        0,
```
Comment (Harkishen-Singh) [inline pkg/tests/testsupport/metric_loader.go:135]: we can remove these

File: pkg/tests/testsupport/metric_loader.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+func (si *sampleIngestor) ingestSamples(ingest IngestFunc) {
+	var wg sync.WaitGroup
+	for i := 0; i < len(si.shards); i++ {
+		go func(shard int) {
+			defer func() {
+				wg.Done()
+			}()
+			var req prompb.WriteRequest = prompb.WriteRequest{Timeseries: make([]prompb.TimeSeries, si.batchSize)}
+			counter := 0
+			for sample := range si.shards[shard] {
```
Comment (Harkishen-Singh) [inline pkg/tests/testsupport/metric_loader.go:223]:
```suggestion
			for ts := range si.shards[shard] {
```
`sample` name is very confusing. This is samples + series, which is proper ts

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
I think we should redesign the PR to have something like a routine sending a batch that contains 1 sample from all series. This will mimic and actual prometheus remote-write, and keep the single iteration of go bench quick.

**Comment by niksajakovljevic** [comment]
> I think we should redesign the PR to have something like a routine sending a batch that contains 1 sample from all series. This will mimic and actual prometheus remote-write, and keep the single iteration of go bench quick.

@Harkishen-Singh I am not sure I follow. Batch size of Prometheus is configurable and so is here. I also didn't get the part about quick bench iteration. Here we run benchmark with 1.5M samples and this data set can be replaced or we can add additional datasets each with different `shape`. Let's discuss more offline.

File: pkg/tests/end_to_end_tests/metric_ingest_bench_test.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+		if err := loader.Close(); err != nil {
+			t.Fatal(err)
+		}
+	}()
+	it := loader.Iterator()
+	sampleCounter := 0
+	for it.Next() {
+		if sampleCounter == 10 {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply pkg/tests/end_to_end_tests/metric_ingest_bench_test.go:35]: That's not the same code. I guess you meant: `for it.Next() && sampleCounter < 10`

File: pkg/tests/end_to_end_tests/metric_ingest_bench_test.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	require.NoError(b, err)
+	defer func() {
+		if err := loader.Close(); err != nil {
+			b.Fatal(err)
+		}
+	}()
+
+	sampleLoader := testsupport.NewSampleIngestor(10, loader.Iterator(), 2000, 0)
+
+	withDB(b, "bench_metric_ingest", func(db *pgxpool.Pool, t testing.TB) {
```
Comment (Harkishen-Singh) [inline pkg/tests/end_to_end_tests/metric_ingest_bench_test.go:61]:
```suggestion
	withDB(b, "bench_e2e_metric_ingest", func(db *pgxpool.Pool, t testing.TB) {
```
Let's be explicit that this benchmark is e2e.

**Review by Harkishen-Singh — APPROVED** [review]

File: pkg/tests/testsupport/metric_loader.go (hunk 4 above)
Comment (antekresic) [reply pkg/tests/testsupport/metric_loader.go:170]: Have you benchmarked this? :grin:

**Review by antekresic — APPROVED** [review]
:tada:

**Pull request merged**